Closure script revised with simpler example (#27)
pupilfirst/wd201-scripts: nodejs-deep-dive/closures-in-js/README.md

# Text
In this lesson, you will learn a very key concept of JavaScript, the **Closures**. 

> I will explain in clear terms what **closure** is, and how you can easily relate it to real-world.

Reviewer: ```suggestion
```

> I guess you haven't heard of this term *closure* previously, and frankly it is not very much self-explanatory as well. From today onwards, you can remember it with a fairly simple term (or synonym) - Legacy! Why, let's find out.

Reviewer: ```suggestion
You probably haven't heard of the term *closure* previously, and it is not very self-explanatory either. From today onwards, you can remember it with a reasonably simple term (or synonym) - Legacy! Why? Let's find out.
```

### What is a closure?
A closure is a feature in JavaScript, where an inner function has access to the outer (enclosing) function’s variables — a scope chain.

The closure has three scope chains:

- it has access to its own scope — that means the variables defined between its curly brackets
- it has access to the outer function’s variables
- it has access to the global variables

Let's understand with a simple example:
```js
function parent() {
    var house = 'Home Sweet Home';
   
    function child() {   
        var mobile = 'iPhone 13'; 
        console.log('I have: ', house, mobile);
    }
   
    return child;
}
var legacy = parent();
console.log(typeof(legacy)); //legacy is of type function

> legacy(); // Output: I have: Home Sweet Home iPhone 13

Reviewer: The example seems contrived.. I know it is a hard topic to write a simple example. But wouldn't a greet function be enough?

```js
function generateGreetings(name) {
  function spanish() {
    console.log(`Hola ${name}!`);
  }
  function english() {
    console.log(`Hello ${name}!`);
  }
  return {spanish, english};
};

const name = 'John';
const greetings = generateGreetings(name);

console.log(typeof(greetings.spanish)); // function

greetings.spanish(); // Hola John!
greetings.english(); // Hello John!
```
